File: tests/zenvia/whatsapp-email.test.ts
import { beforeEach, describe, expect, it, vi } from "vitest";
import type { FastifyInstance } from "fastify";

process.env.DB_PATH = ":memory:";
process.env.MIGRATIONS_DIR = "db/migrations";

async function freshServer(): Promise<FastifyInstance> {
  vi.resetModules();
  const { buildServer } = await import("../../src/server.js");
  return buildServer();
}

describe("POST /zenvia/channels/whatsapp/messages", () => {
  let app: FastifyInstance;

  beforeEach(async () => {
    app = await freshServer();
  });

  it("sends a WhatsApp text message", async () => {
    const res = await app.inject({
      method: "POST",
      url: "/zenvia/channels/whatsapp/messages",
      headers: { "x-api-token": "test-token" },
      payload: {
        from: "whatsapp-account",
        to: "55119999999999",
        contents: [{ type: "text", text: "Hello via WhatsApp" }],
      },
    });

    expect(res.statusCode).toBe(200);
    const body = res.json();
    expect(body.channel).toBe("whatsapp");
    expect(body.contents).toEqual([{ type: "text", text: "Hello via WhatsApp" }]);
  });

  it("sends a WhatsApp file message with idRef/contentRef", async () => {
    const res = await app.inject({
      method: "POST",
      url: "/zenvia/channels/whatsapp/messages",
      headers: { "x-api-token": "test-token" },
      payload: {
        from: "whatsapp-account",
        to: "55119999999999",
        idRef: "7390113b-e120-41b5-8a07-c4567726abc2",
        contentRef: 0,
        contents: [{ type: "file", fileUrl: "https://example.com/file.pdf", fileCaption: "Prescription" }],
      },
    });

    expect(res.statusCode).toBe(200);
    expect(res.json().contents[0].fileUrl).toBe("https://example.com/file.pdf");
  });

  it("rejects an unknown content type", async () => {
    const res = await app.inject({
      method: "POST",
      url: "/zenvia/channels/whatsapp/messages",
      headers: { "x-api-token": "test-token" },
      payload: {
        from: "whatsapp-account",
        to: "55119999999999",
        contents: [{ type: "sticker", stickerId: "abc" }],
      },
    });

    expect(res.statusCode).toBe(400);
  });
});

describe("POST /zenvia/channels/email/messages", () => {
  let app: FastifyInstance;

  beforeEach(async () => {
    app = await freshServer();
  });

  it("sends an email message", async () => {
    const res = await app.inject({
      method: "POST",
      url: "/zenvia/channels/email/messages",
      headers: { "x-api-token": "test-token" },
      payload: {
        from: "no-reply@example.com",
        to: "patient@example.com",
        representative: { name: "GolgiMed" },
        contents: [{ type: "email", subject: "Your results", html: "<b>Hi!</b>" }],
      },
    });

    expect(res.statusCode).toBe(200);
    const body = res.json();
    expect(body.channel).toBe("email");
    expect(body.contents[0].subject).toBe("Your results");
  });

  it("sends an email template message", async () => {
    const res = await app.inject({
      method: "POST",
      url: "/zenvia/channels/email/messages",
      headers: { "x-api-token": "test-token" },
      payload: {
        from: "no-reply@example.com",
        to: "patient@example.com",
        contents: [{ type: "template", templateId: "welcome-email", fields: { name: "Alex" } }],
      },
    });

    expect(res.statusCode).toBe(200);
  });

  it("rejects an email content missing subject", async () => {
    const res = await app.inject({
      method: "POST",
      url: "/zenvia/channels/email/messages",
      headers: { "x-api-token": "test-token" },
      payload: {
        from: "no-reply@example.com",
        to: "patient@example.com",
        contents: [{ type: "email", html: "<b>Hi!</b>" }],
      },
    });

    expect(res.statusCode).toBe(400);
  });
});

File: tests/zenvia/whatsapp_email_test.go
package zenvia_test

import (
	"testing"
	"time"

	"github.com/golgimed/mimic/internal/testutil"
)

func TestWhatsappTextMessage(t *testing.T) {
	app := testutil.New(t, 2*time.Second)

	rec := app.Do(t, "POST", "/zenvia/channels/whatsapp/messages",
		map[string]string{"X-API-TOKEN": "test-token"},
		map[string]any{
			"from":     "whatsapp-account",
			"to":       "55119999999999",
			"contents": []map[string]any{{"type": "text", "text": "Hello via WhatsApp"}},
		},
	)

	if rec.Code != 200 {
		t.Fatalf("expected 200, got %d: %s", rec.Code, rec.Body.String())
	}
	var body map[string]any
	testutil.DecodeJSON(t, rec, &body)
	if body["channel"] != "whatsapp" {
		t.Errorf("channel = %v", body["channel"])
	}
}

func TestWhatsappFileMessage(t *testing.T) {
	app := testutil.New(t, 2*time.Second)

	rec := app.Do(t, "POST", "/zenvia/channels/whatsapp/messages",
		map[string]string{"X-API-TOKEN": "test-token"},
		map[string]any{
			"from":       "whatsapp-account",
			"to":         "55119999999999",
			"idRef":      "7390113b-e120-41b5-8a07-c4567726abc2",
			"contentRef": 0,
			"contents": []map[string]any{
				{"type": "file", "fileUrl": "https://example.com/file.pdf", "fileCaption": "Prescription"},
			},
		},
	)

	if rec.Code != 200 {
		t.Fatalf("expected 200, got %d: %s", rec.Code, rec.Body.String())
	}
	var body struct {
		Contents []map[string]any `json:"contents"`
	}
	testutil.DecodeJSON(t, rec, &body)
	if body.Contents[0]["fileUrl"] != "https://example.com/file.pdf" {
		t.Errorf("fileUrl = %v", body.Contents[0]["fileUrl"])
	}
}

func TestWhatsappRejectsUnknownContentType(t *testing.T) {
	app := testutil.New(t, 2*time.Second)

	rec := app.Do(t, "POST", "/zenvia/channels/whatsapp/messages",
		map[string]string{"X-API-TOKEN": "test-token"},
		map[string]any{
			"from":     "whatsapp-account",
			"to":       "55119999999999",
			"contents": []map[string]any{{"type": "sticker", "stickerId": "abc"}},
		},
	)

	if rec.Code != 400 {
		t.Fatalf("expected 400, got %d", rec.Code)
	}
}

func TestEmailMessage(t *testing.T) {
	app := testutil.New(t, 2*time.Second)

	rec := app.Do(t, "POST", "/zenvia/channels/email/messages",
		map[string]string{"X-API-TOKEN": "test-token"},
		map[string]any{
			"from":           "no-reply@example.com",
			"to":             "patient@example.com",
			"representative": map[string]any{"name": "GolgiMed"},
			"contents":       []map[string]any{{"type": "email", "subject": "Your results", "html": "<b>Hi!</b>"}},
		},
	)

	if rec.Code != 200 {
		t.Fatalf("expected 200, got %d: %s", rec.Code, rec.Body.String())
	}
	var body struct {
		Channel  string           `json:"channel"`
		Contents []map[string]any `json:"contents"`
	}
	testutil.DecodeJSON(t, rec, &body)
	if body.Channel != "email" {
		t.Errorf("channel = %v", body.Channel)
	}
	if body.Contents[0]["subject"] != "Your results" {
		t.Errorf("subject = %v", body.Contents[0]["subject"])
	}
}

func TestEmailTemplateMessage(t *testing.T) {
	app := testutil.New(t, 2*time.Second)

	rec := app.Do(t, "POST", "/zenvia/channels/email/messages",
		map[string]string{"X-API-TOKEN": "test-token"},
		map[string]any{
			"from":     "no-reply@example.com",
			"to":       "patient@example.com",
			"contents": []map[string]any{{"type": "template", "templateId": "welcome-email", "fields": map[string]string{"name": "Alex"}}},
		},
	)

	if rec.Code != 200 {
		t.Fatalf("expected 200, got %d: %s", rec.Code, rec.Body.String())
	}
}

func TestEmailRejectsMissingSubject(t *testing.T) {
	app := testutil.New(t, 2*time.Second)

	rec := app.Do(t, "POST", "/zenvia/channels/email/messages",
		map[string]string{"X-API-TOKEN": "test-token"},
		map[string]any{
			"from":     "no-reply@example.com",
			"to":       "patient@example.com",
			"contents": []map[string]any{{"type": "email", "html": "<b>Hi!</b>"}},
		},
	)

	if rec.Code != 400 {
		t.Fatalf("expected 400, got %d", rec.Code)
	}
}
